api/proto/config/storage.proto

```diff
@@ -242,3 +242,32 @@ message DisksConfig {
   DisksArrayType array_type = 2;
   repeated DisksConfig children = 3;
 }
+
+// SnapshotConfig describes a snapshot for a specific logical
+// volume that must exist on the device. It has a required
+// field volume_uuid for which it was created.
+message SnapshotConfig {
+  // The real snapshot name in ZFS. Filled only in on the EVE side,
+  // at the time the snapshot is created.
+  string uuid = 1;
+
+  // Display name (User-friendly name). This name does not affect
+  // the snapshot properties in ZFS in any way.
+  // Can be filled on the cloud side.
+  // If a snapshot has already been created and this field has changed,
+  // it can be assumed that the friendly name for this snapshot has
+  // been renamed on the cloud side.
+  string display_name = 2;
+
+  // Volume ID of the volume this snapshot belongs to. The field is
+  // required for all messages. Must always be filled in on the
+  // cloud side before sending to create snapshot command.
+  string volume_uuid = 3;
+
+  // The command for the snapshot rollback operation.
+  // Keep in mind that in order to return to any of the
+  // previous snapshots, all intermediate snapshots will be
+  // destroyed in ZFS. These snapshots must be marked as
+  // deleted in the SnapshotInfo message.
```

Comment:
Yes, there is such a situation, but I do not plan to store anything in files, since this makes no sense. ZFS already stores information about snapshots.

There is a certain logic to creating new snapshots.
1) If the config has an empty ID field, then we check for the presence of a snapshot in ZFS, if there is no such snapshot in ZFS, then it is definitely new, and we create it.
2) If Config doesn't exist, check for a snapshot in ZFS, snapshot already exists, add it to EVE.
Even if it didn’t come to us from the cloud, we still need to send it to the cloud so that we can do something with it.
3) The config exists and the ID field is not empty, we check for the presence of a snapshot in ZFS, if this snapshot is not in ZFS, then the snapshot has been deleted, mark it as deleted.

After the RollBack operation, it is also necessary to ask ZFS about all snapshots for the logical volume, and if some snapshots are deleted, mark them as deleted and send their statuses.

I also think that it would be logical, on the cloud side, when sending a rollback operation for any snapshot, to mark all snapshots that were created after it as deleted (provided that the rollback was successful)

Reply:
> If the config has an empty ID field
I cannot understand why we expect empty id field here. Id field expected to be non-empty to identify specific object and to have match between config (generated by controller) and info (controller wait for them and distribute data from info messages by id) objects.